[Merged by Bors] - refactor(*): rename `semimodule` to `module`, delete typeclasses `module` and `vector_space`

Vierkantor opened the pull request:
Splitting typeclasses between `semimodule`, `module` and `vector_space` was causing many (small) issues, so why don't we just get rid of this duplication?

This PR contains the following changes:
 * delete the old `module` and `vector_space` synonyms for `semimodule`
 * find and replace all instances of `semimodule` and `vector_space` to `module`
 * (thereby renaming the previous `semimodule` typeclass to `module`)
 * rename `vector_space.dim` to `module.rank` (in preparation for generalizing this definition to all modules)
 * delete a couple `module` instances that have now become redundant

This find-and-replace has been done extremely naïvely, but it seems there were almost no name clashes and no "clbuttic mistakes". I have gone through the full set of changes, finding nothing weird, and I'm fixing any build issues as they come up (I expect less than 10 declarations will cause conflicts).

Zulip discussion: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/module.20from.20semimodule.20over.20a.20ring

A good example of a workaround required by the `module` abbreviation is the [`triv_sq_zero_ext.module` instance](https://github.com/leanprover-community/mathlib/blob/3b8cfdc905c663f3d99acac325c7dff1a0ce744c/src/algebra/triv_sq_zero_ext.lean#L164).

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
Rename `semimodule` to `module`, delete typeclasses `module` and `vector_space`

Zulip discussion: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/module.20from.20semimodule.20over.20a.20ring

Vierkantor commented on src/algebra/free_algebra.lean:
```diff
@@ -284,7 +284,7 @@ definition.
 attribute [irreducible] ι lift
 -- Marking `free_algebra` irreducible makes `ring` instances inaccessible on quotients.
--- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
+-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20module.20typeclass.20lookup/near/212580241
```
```suggestion
-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
```

Vierkantor commented on test/free_algebra.lean:
```diff
@@ -11,7 +11,7 @@ import linear_algebra.clifford_algebra
 
 There is some discussion about this in
-https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
+https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20module.20typeclass.20lookup/near/212580241
```
```suggestion
https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
```

Vierkantor commented on src/linear_algebra/dimension.lean:
```diff
@@ -38,21 +38,21 @@ variables {ι : Type w} {ι' : Type w'} {η : Type u₁'} {φ : η → Type*}
 
 /-- the dimension of a vector space, defined as a term of type `cardinal` -/
-def vector_space.dim : cardinal :=
+def module.dim : cardinal :=
```
This definition can stay in the `vector_space` namespace even though there is no definition called `vector_space` anymore, right?

```suggestion
def vector_space.dim : cardinal :=
```

Vierkantor commented on src/algebra/module/basic.lean:
```diff
@@ -13,24 +13,14 @@ import algebra.smul_with_zero
 ## Tags
 
-semimodule, module, vector space
+module, module, vector space
```
```suggestion
semimodule, module, vector space
```

Vierkantor commented on src/algebra/module/basic.lean:
```diff
@@ -13,24 +13,14 @@ import algebra.smul_with_zero
+* `module R M` : an additive commutative monoid `M` is a `module` over a
   `semiring R` if for `r : R` and `x : M` their "scalar multiplication `r • x : M` is defined, and
   the operation `•` satisfies some natural associativity and distributivity axioms similar to those
   on a ring.
```
```suggestion
  on a ring.
  
  ## Implementation notes
  In typical mathematical usage, our definition of `module` corresponds to "semimodule", and the word
  "module" is reserved for `module R M` where `R` is a `ring` and `M` an `add_comm_group`. If `R` is
  a `field` and `M` an `add_comm_group`, `M` would be called an `R`-vector space. Since those
  assumptions can be made by changing the typeclasses applied to `R` and `M`, without changing the
  axioms in `module`, mathlib calls everything a `module`.
  
  In older versions of mathlib, we had separate `semimodule` and `vector_space` abbreviations.
  This caused inference issues in some cases, while not providing any real advantages, so we decided
  to use a canonical `module` typeclass throughout.
```

Vierkantor commented on src/linear_algebra/dual.lean:
```diff
@@ -250,11 +250,11 @@ end
 end is_basis
 
-namespace vector_space
+namespace module
```
Same here: I think this can stay the `vector_space` namespace.

Vierkantor commented on src/linear_algebra/exterior_algebra.lean:
```diff
@@ -138,7 +138,7 @@ end
 attribute [irreducible] ι lift
 -- Marking `exterior_algebra` irreducible makes our `ring` instances inaccessible.
--- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
+-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20module.20typeclass.20lookup/near/212580241
```
```suggestion
-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
```

Vierkantor commented on src/linear_algebra/linear_independent.lean:
```diff
@@ -982,7 +982,7 @@ lemma linear_independent_fin_snoc {n} {v : fin n → V} :
 
 /-- See `linear_independent.fin_cons'` for an uglier version that works if you
-only have a semimodule. -/
+only have a module. -/
```
```suggestion
only have a module over a semiring. -/
```

Vierkantor commented on src/linear_algebra/linear_independent.lean:
```diff
@@ -865,13 +865,13 @@ end module
 ### Properties which require `division_ring K`
 
-These can be considered generalizations of properties of linear independence in `vector_space`s.
+These can be considered generalizations of properties of linear independence in `module`s.
```
```suggestion
These can be considered generalizations of properties of linear independence in vector spaces.
```

Vierkantor commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -104,7 +104,7 @@ theorem lift_unique {A : Type*} [semiring A] [algebra R A] (f : M →ₗ[R] A)
 
 -- Marking `tensor_algebra` irreducible makes `ring` instances inaccessible on quotients.
--- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
+-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20module.20typeclass.20lookup/near/212580241
```
```suggestion
-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
```

Vierkantor committed:
Rename `vector_space.dim` to `module.rank`

Vierkantor committed:
We don't need this instance any more

Vierkantor committed:
Replace `dim` with `rank` in `finite_dimensional.lean`

Vierkantor committed:
Rename `dim` to `rank` in `dual.lean`

Vierkantor committed:
Rename `dim` to `module.rank` in `finsupp_vector_space.lean`

Vierkantor committed:
Fix `card_fintype` in `field_theory/finite`

Vierkantor committed:
Rename `dim` -> `module.rank` in `linear_algebra/affine_space`

Vierkantor committed:
Fix two wrongly renamed lemmas

Vierkantor committed:
Update docstrings

Vierkantor committed:
Update docstring

Vierkantor committed:
Wrap overlong lines

Vierkantor committed:
For some reason, there's now a goal that doesn't need to be `simp`ed

Vierkantor committed:
Fix timeout

Vierkantor committed:
Rename `dim` to `module.rank`

Vierkantor committed:
Update references in the library docs

sgouezel commented on docs/undergrad.yaml:
```diff
@@ -4,8 +4,8 @@
-    vector space: 'vector_space'
-    product of vector spaces: 'prod.semimodule'
+    vector space: 'module'
+    product of vector spaces: 'prod.module'
```
Isn't that an unusual name? `module.prod` would be more standard.

Vierkantor replied:
It is the `module` instance on the `prod` type, therefore `prod.module`.

sgouezel commented on src/algebra/module/ordered.lean:
```diff
@@ -10,50 +10,50 @@ import algebra.module.prod
+* We choose to define `ordered_module` as a `Prop`-valued mixin, so that it can be
   used for both modules and algebras
   (the axioms for an "ordered algebra" are exactly that the algebra is ordered as a module).
 * To get ordered modules and ordered vector spaces, it suffices to the replace the
```
```suggestion
* To get ordered modules and ordered vector spaces, it suffices to replace the
```

sgouezel commented on src/algebra/module/ordered.lean:
```diff
@@ -10,50 +10,50 @@ import algebra.module.prod
-@[protect_proj, ancestor semimodule]
-class ordered_semimodule (R M : Type*)
-  [ordered_semiring R] [ordered_add_comm_monoid M] [semimodule R M] : Prop :=
+@[protect_proj, ancestor module]
```
```suggestion
@[protect_proj]
```
There is no ancestor, it is a mixin.

sgouezel commented on src/algebra/module/projective.lean:
```diff
@@ -18,7 +18,7 @@ proof that all free modules are projective.
 ## Main definitions
 
-Let `R` be a ring (or a semiring) and let `M` be an `R`-module (or a semimodule).
+Let `R` be a ring (or a semiring) and let `M` be an `R`-module (or a module).
```
```suggestion
Let `R` be a ring (or a semiring) and let `M` be an `R`-module.
```

sgouezel commented on src/algebra/module/projective.lean:
```diff
@@ -39,8 +39,8 @@ from the free R-module on the type M down to M splits. This is more convenient
-Everything works for semirings and semimodules except that apparently
-we don't have free semimodules, so here we stick to rings.
+Everything works for semirings and modules except that apparently
+we don't have free modules, so here we stick to rings.
```
```suggestion
we don't have free modules over semirings, so here we stick to rings.
```

sgouezel commented on src/analysis/normed_space/basic.lean:
```diff
@@ -1142,27 +1142,27 @@ section semi_normed_space
+-- to take precedence over `semiring.to_module` as this leads to instance paths with better
 -- unification properties.
--- see Note[vector space definition] for why we extend `semimodule`.
+-- see Note[vector space definition] for why we extend `module`.
```
```suggestion
```

sgouezel commented on src/analysis/normed_space/basic.lean:
```diff
@@ -1142,27 +1142,27 @@ section semi_normed_space
+-- to take precedence over `semiring.to_module` as this leads to instance paths with better
 -- unification properties.
--- see Note[vector space definition] for why we extend `semimodule`.
+-- see Note[vector space definition] for why we extend `module`.
```
```suggestion
```

sgouezel commented on src/linear_algebra/dimension.lean:
```diff
@@ -12,13 +12,15 @@ import set_theory.cardinal_ordinal
 
-* The dimension of a vector space is defined as `vector_space.dim : cardinal`.
+* The rank of a module is defined as `module.rank : cardinal`.
+  This is currently only defined for vector spaces.
```
```suggestion
  This is currently only defined for vector spaces, i.e., when the base semiring is a field.
```

sgouezel reviewed (commented):
This looks good to me, thanks a lot!
There is only one point about which I am unsure. After your refactor, we have both `module.rank` and `findim`, which don't go well together. Would `finrank` make sense? (And also adding somewhere a docstring saying that dimension is called rank, that would have enough keywords in it to be easy to find by grepping?)

Vierkantor committed:
Update docstrings

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

Vierkantor committed:
Update docstring

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

Vierkantor committed:
Rename `findim` -> `finrank`

We still call keep calling `finite_dimensional` the same thing,
but we might decide to change that later.

Vierkantor committed:
Fix overlong lines

sgouezel committed:
Merge remote-tracking branch 'upstream/master' into semimodule_to_module

sgouezel committed:
merge master
